cli/azd/pkg/templates/gh_source.go

```diff
@@ -0,0 +1,92 @@
+package templates
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools/github"
+)
+
+// newGhTemplateSource creates a new template source from a Github repository.
+func newGhTemplateSource(
+	ctx context.Context, name string, urlArg string, ghCli *github.Cli, console input.Console) (Source, error) {
+	// urlArg validation:
+	// - accepts only URLs with the following format:
+	//  - https://raw.<hostname>/<owner>/<repo>/<branch>/<path>/<file>.json
+	//  - https://<hostname>/<owner>/<repo>/<branch>/<path>/<file>.json
+	//  - https://api.<hostname>/repos/<owner>/<repo>/contents/<path>/<file>.json
```

Comment:
Can we add some extra docs here in when we would expect the different formats of URLs?